### PR TITLE
WEBAPP-562: Test content not found error poi

### PR DESCRIPTION
**Pull request by Taggotty**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/). 
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **WEBAPP**.

### CODE REVIEW

**Comment by Taggotty** [comment]
The changes in FailureSwitcher are just what I think fits better to our mindset to prefer testing functunality. If you don't like it feel free to object :)

File: src/modules/common/components/__tests__/Failure.spec.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// @flow
+
+import React from 'react'
+import { shallow } from 'enzyme'
+
+import { Failure } from '../Failure'
+
+describe('Failure', () => {
+  const mockTranslate = key => key || 'null'
```
Comment (klinzo) [inline src/modules/common/components/__tests__/Failure.spec.js:9]: Why do we need `|| 'null'`?

File: src/modules/common/components/__tests__/FailureSwitcher.spec.js [hunk 2]
```diff
@@ -29,16 +28,30 @@ describe('FailureSwitcher', () => {
       const error = new ContentNotFoundError({ type: 'event', id: '1234', language, city })
       expect(FailureSwitcher.renderContentNotFoundComponent(error)).toMatchSnapshot()
     })
+
+    it('should render a poi not found failure and match snapshot', () => {
+      const error = new ContentNotFoundError({ type: 'poi', id: 'poiId', language, city })
+      expect(FailureSwitcher.renderContentNotFoundComponent(error)).toMatchSnapshot()
```
Comment (klinzo) [inline src/modules/common/components/__tests__/FailureSwitcher.spec.js:34]: Do you think it makes sense to just assert on this method instead of really rendering the FailureSwitcher?

File: src/modules/common/components/__tests__/FailureSwitcher.spec.js [hunk 3]
```diff
@@ -29,16 +28,30 @@ describe('FailureSwitcher', () => {
...
-    )).toMatchSnapshot()
+    const error = new ContentNotFoundError({ type: 'extra', id: 'sprungbrett', city, language })
+    const spy = jest.spyOn(FailureSwitcher, 'renderContentNotFoundComponent')
+
+    const component = mount(<FailureSwitcher error={error} />)
+
+    expect(spy).toHaveBeenCalledWith(error)
+    expect(component.find(Failure)).toHaveLength(1)
+
+    spy.mockRestore()
```
Comment (klinzo) [inline src/modules/common/components/__tests__/FailureSwitcher.spec.js:47]: Isn't this kind of a duplicate of `should render a extra not found failure and match snapshot`?

File: src/modules/common/components/__tests__/FailureSwitcher.spec.js (hunk 2 above)
Comment (klinzo) [reply src/modules/common/components/__tests__/FailureSwitcher.spec.js]: Still relevant

Comment (Taggotty) [reply src/modules/common/components/__tests__/FailureSwitcher.spec.js]: Hmm. It seems github didn't post my answer. I think it makes sense, as the Failure.spec.js should make sure that rendering of the Failure works. FailureSwitcher should just make sure that the right arguments are passed, don't you think?

File: src/modules/common/components/Failure.js [hunk 4]
```diff
@@ -30,14 +30,18 @@ type PropsType = {|
...
+
   render () {
     const { t, errorMessage, goToPath, goToMessage } = this.props
-    const message = t(goToMessage || 'goTo.start')
 
     return <Centered>
       <div>{t(errorMessage)}</div>
       <div><FontAwesomeIcon icon={faFrown} size='5x' /></div>
-      <Link to={goToPath || new I18nRedirectRouteConfig().getRoutePath({})}>{message}</Link>
+      <Link to={goToPath}>{t(goToMessage)}</Link>
```
Comment (maxammann) [inline src/modules/common/components/Failure.js:44]: Why was this removed?

File: src/modules/common/components/__tests__/Failure.spec.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  it('should render a simple failure and match snapshot', () => {
+    const errorMessage = 'error message'
+    const component = shallow(<Failure errorMessage={errorMessage} t={mockTranslate} />)
+
+    expect(component.instance().props).toEqual({
+      errorMessage: errorMessage,
+      t: mockTranslate,
+      ...Failure.defaultProps
+    })
+    expect(component).toMatchSnapshot()
```
Comment (maxammann) [inline src/modules/common/components/__tests__/Failure.spec.js:20]: I think instead of doing a snapshot test here it would be good to just assert whether the expected text exists using: https://enzymejs.github.io/enzyme/docs/api/ReactWrapper/text.html

Or alternatively whether a link exists to the target you expect.

File: src/modules/common/components/__tests__/Failure.spec.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  it('should render a failure with goToPath and goToMessage and match snapshot', () => {
+    const error = {
+      errorMessage: 'error message',
+      goToPath: 'goTo.extras',
+      goToMessage: 'goTo.extras'
+    }
+    const component = shallow(<Failure {...error} t={mockTranslate} />)
+
+    expect(component.instance().props).toEqual({ ...error, t: mockTranslate })
+    expect(component).toMatchSnapshot()
```
Comment (maxammann) [inline src/modules/common/components/__tests__/Failure.spec.js:32]: I think instead of doing a snapshot test here it would be good to just assert whether the expected text exists using: https://enzymejs.github.io/enzyme/docs/api/ReactWrapper/text.html

Or alternatively whether a link exists to the target you expect.

File: src/modules/common/components/__tests__/__snapshots__/FailureSwitcher.spec.js.snap [hunk 7]
```diff
@@ -1,39 +1,49 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (maxammann) [inline src/modules/common/components/__tests__/__snapshots__/FailureSwitcher.spec.js.snap:1]: These snapshots look clean :+1:

File: src/modules/common/components/__tests__/FailureSwitcher.spec.js [hunk 8]
```diff
@@ -1,44 +1,49 @@
...
+    ${'localNewsItem'}  | ${'/augsburg/en/news/local/1'}
+    ${'tunewsItem'}     | ${'/augsburg/en/news/tu-news/1'}
+    ${'extra'}          | ${'sprungbrett'}
+    ${'poi'}            | ${'1234'}
+    `('render $type component not found failure',
+  ({ type, id }) => {
+    it(`should render a ${type} not found failure and match snapshot`, () => {
+      const error = new ContentNotFoundError({ type, id, language, city })
+      const component = FailureSwitcher.renderContentNotFoundComponent(error)
+      expect(component).toMatchSnapshot()
```
Comment (maxammann) [inline src/modules/common/components/__tests__/FailureSwitcher.spec.js:28]: Maybe a `.dive()` would be cool here. I think then the Translate HOC would be gone in the tests.

**Comment by maxammann** [comment]
Examples for non-snapshot tests:

```
it('should show correct text', () => {
  const wrapper = shallow(<MyComponent />);
  expect(wrapper.text().includes('my text')).toBe(true);
});
```

```
const wrapper = shallow(<div><button className='btn btn-primary'>OK</button></div>);
const button = wrapper.find('.btn'); 
expect(button.text()).to.be.eql('OK');
```

File: src/modules/common/components/Failure.js (hunk 4 above)
Comment (Taggotty) [reply src/modules/common/components/Failure.js:44]: Replaced with default properties to test whether the properties rather than the dom elements.

File: src/modules/common/components/__tests__/FailureSwitcher.spec.js (hunk 8 above)
Comment (Taggotty) [reply src/modules/common/components/__tests__/FailureSwitcher.spec.js:27]: does not work, as dive is an enzyme functionality and the component is not rendered with shallow or mount. The current solution definitly looks cleaner, than a render tree (even with shallow)
